File: Advent of Code/Challenge Solutions/Year 2025/ChallengeSolution03.cs
// Task: https://adventofcode.com/2025/day/3

using Advent_of_Code.Shared;
using Advent_of_Code.Shared.Utilities;

namespace Advent_of_Code.Challenge_Solutions.Year_2025;

public class ChallengeSolution03(IConsole console, ISolutionReader<ChallengeSolution03> reader)
    : ChallengeSolution<ChallengeSolution03>(console, reader)
{
    private const int BatteryJoltageDigitCount = 12;

    public override void SolveFirstPart()
    {
        var batteries = Reader.ReadLines();
        long outputJoltage = 0;

        foreach (var battery in batteries)
        {
            var (firstDigit, firstDigitPosition) = battery[0..^1]
                .Select((value, index) => (value, index))
                .MaxBy(digit => digit.value);

            var secondDigit = battery[(firstDigitPosition + 1)..].Max();

            var batteryJoltage = long.Parse($"{firstDigit}{secondDigit}");

            outputJoltage += batteryJoltage;
        }

        Console.WriteLine($"Output Joltage: {outputJoltage}");
    }

    public override void SolveSecondPart()
    {
        var batteries = Reader.ReadLines();
        long outputJoltage = 0;

        foreach (var battery in batteries)
        {
            outputJoltage += ComputeBatteryJoltage(battery, BatteryJoltageDigitCount);
        }

        Console.WriteLine($"Output Joltage: {outputJoltage}");
    }

    private static long ComputeBatteryJoltage(string battery, int digitCount)
    {
        Stack<char> joltageDigits = [];

        for (var i = 0; i < battery.Length; i++)
        {
            var digit = battery[i];

            while (CanPopDigitFromJoltage(joltageDigits, battery.Length, digitCount, i)
                && joltageDigits.Peek() < digit)
            {
                _ = joltageDigits.Pop();
            }

            if (joltageDigits.Count < digitCount)
            {
                joltageDigits.Push(digit);
            }
        }

        return long.Parse(string.Concat(joltageDigits.Reverse()));
    }

    private static bool CanPopDigitFromJoltage(Stack<char> stack, int batteryLength, int digitCount, int currentPosition)
        => stack.Count > Math.Max(0, digitCount + currentPosition - batteryLength);
}

File: Advent of Code.FSharp/Challenge Solutions/Year 2025/ChallengeSolution03.fs
namespace Advent_of_Code.FSharp.Challenge_Solutions.Year_2025

open Advent_of_Code.Shared
open Advent_of_Code.Shared.Utilities

module ChallengeSolution03 =

    type Solution(getInput: unit -> string list, writeLine: string -> unit) =
            
        let computeOutputJoltage (digitCount: int) (banks: string list) =
            let rec computeBatteryJoltage (digitCount: int) (fromIndex: int) (joltage: int64) (battery: char array) : int64 =
                match digitCount with
                | 0 -> joltage
                | _ ->
                    let searchWindow = battery[fromIndex..(battery.Length - digitCount)]
                    let maxIndex, maxJoltage = searchWindow
                                              |> Array.mapi (fun i jolt -> (i, int64 jolt - int64 '0'))
                                              |> Array.maxBy snd
                    computeBatteryJoltage (digitCount - 1) (fromIndex + maxIndex + 1) (joltage * 10L + maxJoltage) battery                    
                
            banks
            |> List.map _.ToCharArray()
            |> List.map (computeBatteryJoltage digitCount 0 0)
            |> List.sum
        
        member _.SolveFirstPart() =
            let banks = getInput()
            let outputJoltage = computeOutputJoltage 2 banks
                
            writeLine $"Output Joltage: {outputJoltage}"

        member _.SolveSecondPart() : unit =
            let banks = getInput()
            let outputJoltage = computeOutputJoltage 12 banks
                
            writeLine $"Output Joltage: {outputJoltage}"

    type ChallengeSolution03(console: IConsole, reader: ISolutionReader<ChallengeSolution03>) =
        inherit ChallengeSolution<ChallengeSolution03>(console, reader)

        let solution = Solution(reader.ReadLines >> List.ofArray, console.WriteLine)

        override this.SolveFirstPart() = solution.SolveFirstPart()
        override this.SolveSecondPart() = solution.SolveSecondPart()
